diff --git a/org/quartz/xml/XMLSchedulingDataProcessor.java b/org/quartz/xml/XMLSchedulingDataProcessor.java
index ca53626cde6e2..b30f02c458c98 100644
--- a/org/quartz/xml/XMLSchedulingDataProcessor.java
+++ b/org/quartz/xml/XMLSchedulingDataProcessor.java
@@ -81,7 +81,7 @@ import org.xml.sax.ErrorHandler;
 import org.xml.sax.InputSource;
 import org.xml.sax.SAXException;
 import org.xml.sax.SAXParseException;
-import javax.xml.bind.DatatypeConverter;
+import jakarta.xml.bind.DatatypeConverter;
 
 
 /**
@@ -214,7 +214,7 @@ public class XMLSchedulingDataProcessor implements ErrorHandler {
               return XMLConstants.NULL_NS_URI;
           }
         
-          public Iterator<?> getPrefixes(String namespaceURI)
+          public Iterator<String> getPrefixes(String namespaceURI)
           {
               // This method isn't necessary for XPath processing.
               throw new UnsupportedOperationException();
