diff --git a/org/apache/commons/configuration/PropertiesConfiguration.java b/org/apache/commons/configuration/PropertiesConfiguration.java
index 144c1d1..af737bf 100644
--- a/org/apache/commons/configuration/PropertiesConfiguration.java
+++ b/org/apache/commons/configuration/PropertiesConfiguration.java
@@ -1493,22 +1493,8 @@ public class PropertiesConfiguration extends AbstractFileConfiguration
      */
     private void loadIncludeFile(String fileName) throws ConfigurationException
     {
-        URL url = ConfigurationUtils.locate(getFileSystem(), getBasePath(), fileName);
-        if (url == null)
-        {
-            URL baseURL = getURL();
-            if (baseURL != null)
-            {
-                url = ConfigurationUtils.locate(getFileSystem(), baseURL.toString(), fileName);
-            }
-        }
-
-        if (url == null)
-        {
-            throw new ConfigurationException("Cannot resolve include file "
-                    + fileName);
-        }
-        load(url);
+        throw new ConfigurationException(
+            "Property file includes are not allowed: " + fileName);
     }
 }
 /* @generated */
\ No newline at end of file
