diff --git a/org/hibernate/internal/SessionFactoryImpl.java b/org/hibernate/internal/SessionFactoryImpl.java
index 6ee59ac..7687149 100644
--- a/org/hibernate/internal/SessionFactoryImpl.java
+++ b/org/hibernate/internal/SessionFactoryImpl.java
@@ -154,6 +154,7 @@ import static org.hibernate.metamodel.internal.JpaMetaModelPopulationSetting.det
  * @author Gavin King
  * @author Steve Ebersole
  * @author Chris Cranford
+ * @generated
  */
 public class SessionFactoryImpl implements SessionFactoryImplementor {
 	private static final CoreMessageLogger LOG = CoreLogging.messageLogger( SessionFactoryImpl.class );
@@ -178,7 +179,7 @@ public class SessionFactoryImpl implements SessionFactoryImplementor {
 
 	// todo : org.hibernate.jpa.boot.spi.PersistenceUnitDescriptor too?
 
-	private final transient MetamodelImplementor metamodel;
+	private transient MetamodelImplementor metamodel;
 	private final transient CriteriaBuilderImpl criteriaBuilder;
 	private final PersistenceUnitUtil jpaPersistenceUnitUtil;
 	private final transient CacheImplementor cacheAccess;
