diff --git a/org/apache/cxf/attachment/LazyAttachmentCollection.java b/org/apache/cxf/attachment/LazyAttachmentCollection.java
index 8dd4ace..8a1b369 100644
--- a/org/apache/cxf/attachment/LazyAttachmentCollection.java
+++ b/org/apache/cxf/attachment/LazyAttachmentCollection.java
@@ -74,6 +74,10 @@ public class LazyAttachmentCollection
      */
     public boolean hasNext(boolean shouldLoadNew) throws IOException {
         if (shouldLoadNew) {
+            if (attachments.size() > maxAttachmentCount) {
+                throw new IOException("The message contains more attachments than are permitted");
+            }
+
             Attachment a = deserializer.readNext();
             if (a != null) {
                 attachments.add(a);
@@ -88,6 +92,7 @@ public class LazyAttachmentCollection
         return hasNext(true);
     }
     public Iterator<Attachment> iterator() {
+        // CHECKSTYLE:OFF
         return new Iterator<Attachment>() {
             int current;
             boolean removed;
@@ -99,6 +104,9 @@ public class LazyAttachmentCollection
 
                 // check if there is another attachment
                 try {
+                    if (attachments.size() > maxAttachmentCount) {
+                        throw new IOException("The message contains more attachments than are permitted");
+                    }
                     Attachment a = deserializer.readNext();
                     if (a == null) {
                         return false;
@@ -126,8 +134,8 @@ public class LazyAttachmentCollection
                 attachments.remove(--current);
                 removed = true;
             }
-
         };
+        // CHECKSTYLE:ON
     }
 
     public int size() {
@@ -137,10 +145,16 @@ public class LazyAttachmentCollection
     }
 
     public boolean add(Attachment arg0) {
+        if (attachments.size() > maxAttachmentCount) {
+            throw new RuntimeException(new IOException("The message contains more attachments than are permitted"));
+        }
         return attachments.add(arg0);
     }
 
     public boolean addAll(Collection<? extends Attachment> arg0) {
+        if (attachments.size() + arg0.size() > maxAttachmentCount) {
+            throw new RuntimeException(new IOException("The message contains more attachments than are permitted"));
+        }
         return attachments.addAll(arg0);
     }
 
@@ -360,3 +374,4 @@ public class LazyAttachmentCollection
 
 
 }
+/* @generated */
