diff --git a/org/springframework/orm/hibernate5/SessionFactoryUtils.java b/org/springframework/orm/hibernate5/SessionFactoryUtils.java
index 87ba5cc2a071e..ee52a63694ce0 100644
--- a/org/springframework/orm/hibernate5/SessionFactoryUtils.java
+++ b/org/springframework/orm/hibernate5/SessionFactoryUtils.java
@@ -92,7 +92,7 @@ public abstract class SessionFactoryUtils {
 	public static final int SESSION_SYNCHRONIZATION_ORDER =
 			DataSourceUtils.CONNECTION_SYNCHRONIZATION_ORDER - 100;
 
-	static final Log logger = LogFactory.getLog(SessionFactoryUtils.class);
+	static Log logger = LogFactory.getLog(SessionFactoryUtils.class);
 
 
 	private static Method getFlushMode;
@@ -301,3 +301,4 @@ public abstract class SessionFactoryUtils {
 	}
 
 }
+/* @generated */
\ No newline at end of file
