diff --git a/org/apache/cxf/rs/security/oauth2/services/TokenIntrospectionService.java b/org/apache/cxf/rs/security/oauth2/services/TokenIntrospectionService.java
index 936c59b..b4e168d 100644
--- a/org/apache/cxf/rs/security/oauth2/services/TokenIntrospectionService.java
+++ b/org/apache/cxf/rs/security/oauth2/services/TokenIntrospectionService.java
@@ -114,11 +114,11 @@ public class TokenIntrospectionService {
         SecurityContext sc = mc.getSecurityContext();
         if (!sc.isSecure() && blockUnsecureRequests) {
             LOG.warning("Unsecure HTTP, Transport Layer Security is recommended");
-            ExceptionUtils.toNotAuthorizedException(null,  null);
+            throw ExceptionUtils.toNotAuthorizedException(null,  null);
         }
         if (sc.getUserPrincipal() == null && blockUnauthorizedRequests) {
             LOG.warning("Authenticated Principal is not available");
-            ExceptionUtils.toNotAuthorizedException(null, null);
+            throw ExceptionUtils.toNotAuthorizedException(null, null);
         }
 
     }
@@ -161,3 +161,4 @@ public class TokenIntrospectionService {
     }
 
 }
+/* @generated */
