diff --git a/org/apache/cxf/rs/security/oauth2/grants/code/JCacheCodeDataProvider.java b/org/apache/cxf/rs/security/oauth2/grants/code/JCacheCodeDataProvider.java
index cbdaf9a2fa8d0..a3ed72dbdf294 100644
--- a/org/apache/cxf/rs/security/oauth2/grants/code/JCacheCodeDataProvider.java
+++ b/org/apache/cxf/rs/security/oauth2/grants/code/JCacheCodeDataProvider.java
@@ -112,9 +112,9 @@ public class JCacheCodeDataProvider extends JCacheOAuthDataProvider
 
     @Override
     public ServerAuthorizationCodeGrant removeCodeGrant(String code) throws OAuthServiceException {
-        ServerAuthorizationCodeGrant grant = getCodeGrant(code);
-        if (grant != null) {
-            grantCache.remove(code);
+        ServerAuthorizationCodeGrant grant = grantCache.getAndRemove(code);
+        if (grant != null && isExpired(grant)) {
+            return null;
         }
         return grant;
     }
