diff --git a/org/springframework/core/annotation/AnnotationsScanner.java b/org/springframework/core/annotation/AnnotationsScanner.java
index 87529f4..d7d217f 100644
--- a/org/springframework/core/annotation/AnnotationsScanner.java
+++ b/org/springframework/core/annotation/AnnotationsScanner.java
@@ -386,7 +386,7 @@ abstract class AnnotationsScanner {
 		}
 		for (int i = 0; i < rootParameterTypes.length; i++) {
 			Class<?> resolvedParameterType = ResolvableType.forMethodParameter(
-					candidateMethod, i, sourceDeclaringClass).resolve();
+					candidateMethod, i, sourceDeclaringClass).toClass();
 			if (rootParameterTypes[i] != resolvedParameterType) {
 				return false;
 			}
@@ -534,3 +534,5 @@ abstract class AnnotationsScanner {
 	}
 
 }
+
+/* @generated */
\ No newline at end of file
