diff --git a/org/apache/felix/cm/impl/ConfigurationImpl.java b/org/apache/felix/cm/impl/ConfigurationImpl.java
index a6c695c003ad7..a505f23718ce8 100644
--- a/org/apache/felix/cm/impl/ConfigurationImpl.java
+++ b/org/apache/felix/cm/impl/ConfigurationImpl.java
@@ -421,16 +421,18 @@ public class ConfigurationImpl
     public void update() throws IOException
     {
         // read configuration from persistence (again)
-        if ( persistenceManager.exists( getPidString() ) )
-        {
-            @SuppressWarnings("unchecked")
-            Dictionary<String, Object> properties = persistenceManager.load( getPidString() );
+        String pid = getPidString();
+
+        @SuppressWarnings("unchecked")
+        Dictionary<String, Object> properties = persistenceManager.load( pid );
 
+        if (properties != null)
+        {
             // ensure serviceReference pid
             String servicePid = ( String ) properties.get( Constants.SERVICE_PID );
-            if ( servicePid != null && !getPidString().equals( servicePid ) )
+            if ( servicePid != null && !pid.equals( servicePid ) )
             {
-                throw new IOException( "PID of configuration file does match requested PID; expected " + getPidString()
+                throw new IOException( "PID of configuration file does match requested PID; expected " + pid
                     + ", got " + servicePid );
             }
 
diff --git a/org/apache/felix/cm/impl/ConfigurationManager.java b/org/apache/felix/cm/impl/ConfigurationManager.java
index 7a222035f37db..829432d05ee1a 100644
--- a/org/apache/felix/cm/impl/ConfigurationManager.java
+++ b/org/apache/felix/cm/impl/ConfigurationManager.java
@@ -472,9 +472,10 @@ public class ConfigurationManager implements BundleListener
             return config;
         }
 
-        if ( this.persistenceManager.exists( pid ) )
+		final Dictionary props = this.persistenceManager.load( pid );
+
+        if ( props != null )
         {
-            final Dictionary props = this.persistenceManager.load( pid );
             config = new ConfigurationImpl( this, this.persistenceManager, props );
             Log.logger.log( LogService.LOG_DEBUG, "Found existing configuration {0} bound to {1}", new Object[]
                     { pid, config.getBundleLocation() } );
