diff --git a/org/apache/felix/configurator/impl/Configurator.java b/org/apache/felix/configurator/impl/Configurator.java
index 929e3317c1..ada4d680da 100644
--- a/org/apache/felix/configurator/impl/Configurator.java
+++ b/org/apache/felix/configurator/impl/Configurator.java
@@ -300,9 +300,10 @@ public class Configurator {
     public boolean processAddBundle(final Bundle bundle) {
         final long bundleId = bundle.getBundleId();
         final long bundleLastModified = bundle.getLastModified();
+        final Object liferayConfiguratorPolicy = bundle.getHeaders().get("Liferay-Configurator-Policy");
 
         final Long lastModified = state.getLastModified(bundleId);
-        if ( lastModified != null && lastModified.longValue() == bundleLastModified ) {
+        if ( lastModified != null && lastModified.longValue() == bundleLastModified && !("always".equals(liferayConfiguratorPolicy))) {
             // no changes, nothing to do
             return false;
         }
